### ISSUE TITLE
SIGSEGV on db open

### ISSUE THREAD

**Issue by Banu27** [facebook/rocksdb#4670]
Rocksdb version: 5.14.2
JVM version: JDK-1_8_0_72

### Actual behavior: Segmentation fault during rocksdb open. This wasn't happening in the past and we have been on this version for almost 3 months now. Please let me know if any other information would be helpful 

Stack trace:

`R14=0x00007f7ce3ffb060: <offset 0x8fa060> in /export/content/lid/apps/followfeed-storage/50b8fc1343aa4fbdcfd954006b616c5e8a73e5ef/tmp/librocksdbjni2249206131915847291.so at 0x00007f7ce3701000
R15=0x000000000000003a is an unknown value

Stack: [0x00007f7ab04f6000,0x00007f7ab05f7000],  sp=0x00007f7ab05f4650,  free space=1017k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [librocksdbjni2249206131915847291.so+0x4435d5]  rocksdb::InstrumentedMutex::Lock()+0x125
C  [librocksdbjni2249206131915847291.so+0x382962]  rocksdb::DBImpl::Open(rocksdb::DBOptions const&, std::string const&, std::vector<rocksdb::ColumnFamilyDescriptor, std::allocator<rocksdb::ColumnFamilyDescriptor> > const&, std::vector<rocksdb::ColumnFamilyHandle*, std::allocator<rocksdb::ColumnFamilyHandle*> >*, rocksdb::DB**, bool)+0x9b2
C  [librocksdbjni2249206131915847291.so+0x383c58]  rocksdb::DB::Open(rocksdb::DBOptions const&, std::string const&, std::vector<rocksdb::ColumnFamilyDescriptor, std::allocator<rocksdb::ColumnFamilyDescriptor> > const&, std::vector<rocksdb::ColumnFamilyHandle*, std::allocator<rocksdb::ColumnFamilyHandle*> >*, rocksdb::DB**)+0x18
C  [librocksdbjni2249206131915847291.so+0x384322]  rocksdb::DB::Open(rocksdb::Options const&, std::string const&, rocksdb::DB**)+0x6b2
C  [librocksdbjni2249206131915847291.so+0x2db03a]  std::_Function_handler<rocksdb::Status ()(rocksdb::Options const&, std::string const&, rocksdb::DB**), rocksdb::Status (*)(rocksdb::Options const&, std::string const&, rocksdb::DB**)>::_M_invoke(std::_Any_data const&, rocksdb::Options const&, std::string const&, rocksdb::DB**)+0x1a
C  [librocksdbjni2249206131915847291.so+0x2d32fe]  rocksdb_open_helper(JNIEnv_*, long, _jstring*, std::function<rocksdb::Status ()(rocksdb::Options const&, std::string const&, rocksdb::DB**)>)+0x7e
C  [librocksdbjni2249206131915847291.so+0x2d343e]  Java_org_rocksdb_RocksDB_open__JLjava_lang_String_2+0x3e
j  org.rocksdb.RocksDB.open(JLjava/lang/String;)J+0
J 15025 C1 com.linkedin.utils.rocksdb.RocksDBAccessor.open(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Lcom/linkedin/utils/rocksdb/RocksDbConfig;Lcom/linkedin/followfeed/publicapis/metrics/Sensor;)Lcom/linkedin/utils/rocksdb/RocksDBAccessor; (181 bytes) @ 0x00007f96a3dc59b4 [0x00007f96a3dc4f60+0xa54]
j  com.linkedin.utils.rocksdb.RocksDBStorageImpl.lambda$openDb$1(I)V+101
j  com.linkedin.utils.rocksdb.RocksDBStorageImpl$$Lambda$131.run()V+8
J 14343 C1 java.util.concurrent.Executors$RunnableAdapter.call()Ljava/lang/Object; (14 bytes) @ 0x00007f96a1630a0c [0x00007f96a1630900+0x10c]
J 14346 C1 java.util.concurrent.FutureTask.run()V (126 bytes) @ 0x00007f96a365f784 [0x00007f96a365f540+0x244]
J 14343 C1 java.util.concurrent.Executors$RunnableAdapter.call()Ljava/lang/Object; (14 bytes) @ 0x00007f96a1630a0c [0x00007f96a1630900+0x10c]
J 14346 C1 java.util.concurrent.FutureTask.run()V (126 bytes) @ 0x00007f96a365f784 [0x00007f96a365f540+0x244]
J 14676 C1 java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V (225 bytes) @ 0x00007f96a3c91abc [0x00007f96a3c90ae0+0xfdc]
J 14461 C1 java.util.concurrent.ThreadPoolExecutor$Worker.run()V (9 bytes) @ 0x00007f96a13a6c04 [0x00007f96a13a6b00+0x104]
J 14358 C1 java.lang.Thread.run()V (17 bytes) @ 0x00007f96a14bdabc [0x00007f96a14bd980+0x13c]
v  ~StubRoutines::call_stub
V  [libjvm.so+0x68c616]  JavaCalls::call_helper(JavaValue*, methodHandle*, JavaCallArguments*, Thread*)+0x1056
V  [libjvm.so+0x68cb21]  JavaCalls::call_virtual(JavaValue*, KlassHandle, Symbol*, Symbol*, JavaCallArguments*, Thread*)+0x321
V  [libjvm.so+0x68cfc7]  JavaCalls::call_virtual(JavaValue*, Handle, KlassHandle, Symbol*, Symbol*, Thread*)+0x47
V  [libjvm.so+0x723d80]  thread_entry(JavaThread*, Thread*)+0xa0
V  [libjvm.so+0xa69dcf]  JavaThread::thread_main_inner()+0xdf
V  [libjvm.so+0xa69efc]  JavaThread::run()+0x11c
V  [libjvm.so+0x91d9d8]  java_start(Thread*)+0x108
C  [libpthread.so.0+0x7dd5]  start_thread+0xc5
`

**Comment by sagar0** [comment]
If you are able to consistently reproduce the issue, could you provide a small test case? That would make things much easier to debug.

**Comment by Banu27** [comment]
Hi @sagar0 ,

We are not able to reproduce the crash at this point but it did happen about 5/6 times in the past week. Also, we have unit tests for the db open codepath and they are passing fine. Since there is no definitive set of steps to reproduce, Could you maybe publish a version on maven which enables c++ line numbers? We can pull it from there and have it run in production. When the failure happens, we will have the logs for it.

**Comment by sagar0** [comment]
@Banu27 -- I see a few places in DBImpl::Open where InstrumentedMutex::Lock() is called, so it would be quite difficult to debug without having the line numbers and/or a reproducible test case. 

> Could you maybe publish a version on maven which enables c++ line numbers?

We don't publish debug builds to maven; moreover the debug builds will have much worse performance to run in production. However I think its worth considering publishing Debug jars to maven in the future. @adamretter what do you think?

**Comment by adamretter** [comment]
@sagar0 we could publish a debug version to Maven Central as well if it is desirable. Let me know...

**Comment by guillaumepitel** [comment]
It may be related to #4773 : the problem also occurs in a Mutex.Lock()

The problem doesn't appear in DEBUG mode, though. But it's reproducible

**Comment by adamretter** [comment]
@sagar0 @ajkr Any further thoughts on whether you also want a debug version published?

**Comment by adamretter** [comment]
@jay-zhuang @ramvadiv What do you think about us also publishing debug binaries to Maven Central?

**Comment by jay-zhuang** [comment]
will the debug jar binaries give the user more information when there's an exception? Or what's the other benefit?
Typically, java doesn't have debug and release version, no sure having a debug version would confuse the java developer.

**Comment by adamretter** [comment]
@jay-zhuang Yes. The native part of the debug binaries are compiled with `DEBUG_LEVEL=2` which means that the full-unmanged stacktrace is available on SIGSEGV and line numbers are resolvable to source code.

javac does have options for producing code which is stripped of debug information for when you need the absolute most performance possible, however we don't ship the RocksJava code like that (at the moment).

**Comment by jay-zhuang** [comment]
> @jay-zhuang Yes. The native part of the debug binaries can be compiled with `DEBUG_LEVEL=2` which means that the full-unmangled stacktrace is available on SIGSEGV and line numbers are resolvable to source code.

If that's the case, I could see the value of having a debug version, especially right now, the java exception doesn't give us too much information about what's wrong inside of the c++ code.
